org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/plain/GetProjectSourceLocationsCommand.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.jdt.ls.extension.core.internal.plain;
+
+import static org.eclipse.che.jdt.ls.extension.core.internal.Utils.ensureNotCancelled;
+
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.List;
+import org.eclipse.che.jdt.ls.extension.core.internal.JavaModelUtil;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.IPackageFragment;
+import org.eclipse.jdt.core.IPackageFragmentRoot;
+import org.eclipse.jdt.ls.core.internal.JavaLanguageServerPlugin;
+
+/**
+ * The command to get all available source folder paths.
+ *
+ * @author V. Rubezhny
+ */
+public class GetProjectSourceLocationsCommand {
+  /**
+   * Find all possible destination source folder paths.
+   *
+   * @return list of the destination source folder paths
+   */
+  public static List<String> execute(List<Object> arguments, IProgressMonitor pm) {
+    ensureNotCancelled(pm);
+    validateArguments(arguments);
+
+    final String projectUri = (String) arguments.get(0);
+    IJavaProject javaProject = JavaModelUtil.getJavaProject(projectUri);
+
+    List<String> result = new ArrayList<>();
+    if (javaProject.exists()) {
+      try {
+        result.addAll(getAllPackageFragmentPaths(javaProject));
+      } catch (CoreException e) {
+        JavaLanguageServerPlugin.logException(e.getMessage(), e);
+      }
+    }
+    return result;
+  }
+
+  private static List<String> getAllPackageFragmentPaths(IJavaProject javaProject)
+      throws CoreException {
+    List<String> result = new ArrayList<>();
+    for (IPackageFragmentRoot packageFragmentRoot : javaProject.getAllPackageFragmentRoots()) {
+      if (packageFragmentRoot.getKind() == IPackageFragmentRoot.K_SOURCE
+          && javaProject.getPath().isPrefixOf(packageFragmentRoot.getPath())) {
+        result.addAll(getPackageFragmentPaths(packageFragmentRoot));
+      }
+    }
+    return result;
+  }
+
+  private static List<String> getPackageFragmentPaths(IPackageFragmentRoot packageFragmentRoot)
+      throws CoreException {
+    List<String> result = new ArrayList<>();
+    IJavaElement[] children = packageFragmentRoot.getChildren();
+    if (children != null) {
+      for (IJavaElement child : children) {
+        if (child instanceof IPackageFragment) {
+          result.add(
```

Comment:
Why not just use child.getResource().getLocationURI() here? (with null checks)

Reply:
Yes, that's much easier. However, it is still needed to convert URI string in form of "file:///..."instead of "file:/...".